### PR TITLE
Add support for metaplex fungible token metadata

### PR DESCRIPTION
**Pull request by ThornM9**
Referencing issue #1383

### CODE REVIEW

File: packages/recoil/src/atoms/solana/token.tsx [hunk 1]
```diff
@@ -106,20 +111,61 @@ export const solanaTokenNativeBalance = selectorFamily<
     (tokenAddress: string) =>
     ({ get }: any) => {
       const tokenAccount = get(solanaTokenAccountsMap({ tokenAddress }));
-      if (!tokenAccount) {
```
Comment (armaniferrante) [inline packages/recoil/src/atoms/solana/token.tsx:109]: Why was this removed? Seems like it could be necessary still.

File: packages/recoil/src/atoms/solana/token.tsx [hunk 2]
```diff
@@ -106,20 +111,61 @@ export const solanaTokenNativeBalance = selectorFamily<
...
+      const tokenMetadata: TokenMetadata | null = splTokenMetadata.find(
+        (m: TokenMetadataString) =>
+          metadataAddress.equals(new PublicKey(m.publicKey))
+      );
+      if (tokenMetadata) {
+        const displayBalance = ethers.utils.formatUnits(
+          nativeBalance,
+          tokenMetadata.decimals
+        );
+        const { symbol: ticker, image: logo, name } = tokenMetadata.uriMetadata;
```
Comment (tomlinton) [inline packages/recoil/src/atoms/solana/token.tsx:137]: This is not guaranteed to be defined I think?

Comment (ThornM9) [reply packages/recoil/src/atoms/solana/token.tsx]: True, updated that

File: packages/recoil/src/atoms/solana/token.tsx (hunk 1 above)
Comment (ThornM9) [reply packages/recoil/src/atoms/solana/token.tsx:109]: Fixed

**Comment by armaniferrante** [comment]
this has regressed the token list 

Screen Shot 2022-11-13 at 8 40 19 AM

**Comment by ThornM9** [comment]
@armaniferrante that's fixed now, didn't realise some token registry tokens had a empty metaplex metadata account. Behaviour now is to only use metadata if it's defined and token registry is empty.

**Comment by ThornM9** [comment]
@armaniferrante fixed, didn't realise that some token registry tokens would also have a metaplex metadata account. Behaviour is now to only use metaplex metadata if there's no token registry.

File: packages/common/src/solana/programs/token.ts [hunk 3]
```diff
@@ -105,14 +131,92 @@ export async function customSplTokenAccounts(
...
+  const metadataArr = Array.from(tokens.values());
+
+  // if token standard is available then use it. otherwise determine fungibility by decimals
+  const tokensWithMetadata = metadataArr
+    .map((m) => {
+      return { ...m, decimals: m.decimals || 0 };
+    })
+    .filter((m) =>
+      m.metadata.tokenStandard
+        ? m.metadata.tokenStandard == TokenStandard.FungibleAsset ||
```
Comment (tomlinton) [inline packages/common/src/solana/programs/token.ts:189]: I think FungibleAsset should end up in the NFT view rather than grouped with SPL tokens, at least that is the guidance here https://docs.metaplex.com/programs/token-metadata/token-standard#introduction

File: packages/common/src/solana/programs/token.ts [hunk 4]
```diff
@@ -73,26 +78,47 @@ export async function customSplTokenAccounts(
...
+    tokenAccountsArray,
+    tokenAccountToMint
   );
 
   //
-  // Fetch the metadata uri and interpert as NFTs.
+  // Fetch all metadata uris
   //
-  const nftMetadata = await fetchSplMetadataUri(
+  const tokenAccountsToMetadataWithUri = await fetchSplMetadataUri(
```
Comment (tomlinton) [inline packages/common/src/solana/programs/token.ts:99]: Tokens that metadata could not be fetched for are effectively filtered out of this array from the `fetchSplMetadataUri` function, that means they aren't separated into NFTs below. This is mostly spam NFTs, but we should try and separate them into NFTs correctly anyway. This will stop them showing up in the token list and provide the right access to burning options in the NFT detail pages.

Screenshot 2022-11-15 at 1 53 03 PM

**Comment by tomlinton** [comment]
Left a few more comments @ThornM9, appreciate your effort here. There is a lot of edge cases with spammy tokens.

**Comment by tomlinton** [comment]
This closes https://github.com/coral-xyz/backpack/issues/1352

**Comment by tomlinton** [comment]
@ThornM9 will you be finishing this up? Happy to complete it for you if you are busy, but would like to get it in soon given some refactoring thats going to happen shortly.

**Comment by ThornM9** [comment]
@tomlinton Really appreciate the feedback, would like to finish it off but been pretty busy. If it's holding you up, might be better if you take it up. Otherwise I'll probably be able to get to it early next week.

**Comment by armaniferrante** [comment]
Completed here https://github.com/coral-xyz/backpack/pull/1828. Thank you for this! Was very helpful as a starting point!
